src/rust/ensogl/lib/components/src/scrollbar.rs

```diff
@@ -0,0 +1,267 @@
+            frp.source.thumb_position <+ thumb_position.value;
+
+
+            // === Mouse position in local coordinates ===
+
+            mouse_position <- mouse.position.map(f!([scene,model](pos)
+                scene.screen_to_object_space(&model,*pos)));
+
+            // We will initialize the mouse position with `Vector2(f32::NAN,f32::NAN)`, because the
+            // default `Vector2(0.0,0.0)` would reveal the scrollbar before we get the actual mouse
+            // coordinates.
+            init_mouse_position <- source::<Vector2>();
+            mouse_position      <- any(&mouse_position,&init_mouse_position);
+
+
+            // === Color ===
+
+            init_color    <- any_mut::<()>();
+            default_color <- all(&default_color,&init_color)._0().map(|c| color::Lch::from(*c));
+            hover_color   <- all(&hover_color,&init_color)._0().map(|c| color::Lch::from(*c));
+
+            engaged                  <- base_frp.track_hover || base_frp.is_dragging_track;
+            thumb_color.target_color <+ engaged.switch(&default_color,&hover_color);
+            eval thumb_color.value((c) model.set_track_color(color::Rgba::from(*c)));
+
+
+            // === Hiding ===
+
+            // We start a delayed animation whenever the bar is scrolled to a new place (it is
+            // active). This will instantly reveal the scrollbar and hide it after the delay has
+            // passed.
+            activity_cool_off.frp.reset <+ active;
+            activity_cool_off.frp.start <+ active;
+
+            recently_active <- bool(&activity_cool_off.frp.on_end,&activity_cool_off.frp.on_reset);
+
+            // The signed distance between the cursor and the edge of the scrollbar. If the cursor
+            // is further left or right than the ends of the scrollbar then we count the distance as
+            // infinite.
+            vert_mouse_distance <- all_with(&mouse_position,&frp.set_length,|&pos,&length| {
+                let scrollbar_x_range = (-length/2.0)..=(length/2.0);
+                if scrollbar_x_range.contains(&pos.x) {
+                    pos.y.abs() - WIDTH / 2.0
+                } else {
+                    f32::INFINITY
+                }
+            });
+
+            thumb_color.target_alpha <+ all_with5(&recently_active,&base_frp.is_dragging_track,
+                &vert_mouse_distance,&frp.set_thumb_size,&frp.set_max,
+                |&active,&dragging,&dist,&thumb_size,&max| {
+                    let thumb_fills_bar = thumb_size >= max;
+                    if thumb_fills_bar {
+                        0.0
+                    } else if active || dragging {
+                        1.0
+                    } else {
+                        if dist <= 0.0 {
+                            1.0
+                        } else {
+                            (0.7 - dist / 20.0).max(0.0).min(1.0)
+                        }
+                    }
```

Comment:
it deserves its own function. And maybe a bit of explanation of the 0.7 - dist / 20.0 magic.